src/Algebra/Action/Bundles.agda

```diff
@@ -0,0 +1,141 @@
+-- The Agda standard library
+--
+-- Setoid Actions and Monoid Actions
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Algebra.Action.Bundles where
+
+open import Algebra.Action.Structures.Raw using (IsRawLeftAction; IsRawRightAction)
+open import Algebra.Bundles using (Monoid)
+
+open import Data.List.Base using ([]; _++_)
+import Data.List.Relation.Binary.Equality.Setoid as ≋
+open import Data.Product.Base using (curry; uncurry)
+open import Data.Product.Relation.Binary.Pointwise.NonDependent using (_×ₛ_)
+
+open import Function.Bundles using (Func)
+
+open import Level using (Level; _⊔_)
+
+open import Relation.Binary.Bundles using (Setoid)
+
+private
+  variable
+    a c r ℓ : Level
+
+
+------------------------------------------------------------------------
+-- Basic definition: a Setoid action yields underlying raw action
+
+module SetoidAction (M : Setoid c ℓ) (A : Setoid a r) where
+
+  private
+
+    open module M = Setoid M using ()
+    open module A = Setoid A using (_≈_)
+
+  record Left : Set (a ⊔ r ⊔ c ⊔ ℓ) where
+
+    field
+      act : Func (M ×ₛ A) A
+
+    isRawLeftAction : IsRawLeftAction M._≈_ _≈_
+    isRawLeftAction = record { _ᴹ∙ᴬ_ = curry to ; ∙-cong = curry cong }
+      where open Func act
+
+  record Right : Set (a ⊔ r ⊔ c ⊔ ℓ) where
+
+    field
+      act : Func (A ×ₛ M) A
+
+    isRawRightAction : IsRawRightAction M._≈_ _≈_
+    isRawRightAction = record { _ᴬ∙ᴹ_ = curry to ; ∙-cong = curry cong }
+      where open Func act
+
+
+------------------------------------------------------------------------
+-- A Setoid action yields an iterated List action
+
+module _ {M : Setoid c ℓ} {A : Setoid a r} where
+
+  open SetoidAction
+
+  open ≋ M using (≋-setoid)
+
+  leftListAction : (leftAction : Left M A) → Left ≋-setoid A
+  leftListAction leftAction = record
+    { act = record
+      { to = uncurry _ᴹ⋆ᴬ_ ; cong = uncurry ⋆-cong }
+    }
+    where open Left leftAction; open IsRawLeftAction isRawLeftAction
+
+  rightListAction : (rightAction : Right M A) → Right ≋-setoid A
+  rightListAction rightAction = record
+    { act = record
+      { to = uncurry _ᴬ⋆ᴹ_ ; cong = uncurry ⋆-cong }
+    }
+    where open Right rightAction; open IsRawRightAction isRawRightAction
+
+
+------------------------------------------------------------------------
+-- Definition: indexed over an underlying raw action
+
+module MonoidAction (M : Monoid c ℓ) (A : Setoid a r) where
+
+  private
+
+    module M = Monoid M
+    module A = Setoid A
+    open ≋ M.setoid using (≋-refl)
+
+  record Left (leftAction : SetoidAction.Left M.setoid A) : Set (a ⊔ r ⊔ c ⊔ ℓ)
+    where
+
+    open SetoidAction.Left leftAction public
+      using (isRawLeftAction)
+    open IsRawLeftAction isRawLeftAction public
+      renaming (_ᴹ∙ᴬ_ to _∙_; _ᴹ⋆ᴬ_ to _⋆_)
+
+    field
+      ∙-act  : ∀ m n x → m M.∙ n ∙ x A.≈ m ∙ n ∙ x
```

Comment:
Ah, OK, I was thinking that they were 'noise', and had tuned the infix decls. to ensure I didn't need them, (the `M.` qualification is the clue we're switching type...). Sigh.

Reply:
I suppose I could eschew the renaming, and then the 'bracketing' might be clearer?

Reply:
I ended up making different choices. Maybe not to your taste, either...!?

Reply:
(Some of) My colleagues in WG2.1 would certainly agree with you as regards 'asymmetric symbols for asymmetric concepts', and even I do, to an extent. But *mathematically*, I find it compelling to have the pun between the action map and the monoid map, despite railing elsewhere against puns as being harmful. My 'ugly' superscripting is a way of introducing asymmetry, but then overturning it in the service of 'ordinary practice'.

As to the RHS with the 'iterated' action being given a named definition, is that (not)... overkill?

I think these naming/notation questions are tricky, and to some extent governed by convention (`List.cons` being the worst offender?), and for sure I think it's a bigger discussion than concerning this PR. Happy to put it out to (wider) discussion, but atm I'm reluctant to change... neither do I want this to be a hill on which to die!

Sigh.

Reply:
By 'named definition' I meant like `Commutative` and `Associative` -- maybe that wasn't clear? My feeling is that if we looked, we'd find uses of something equivalent to this action structure already in the library!

Mathematical puns are cute 'in the small' but my experience is that they don't scale well. Actions are spreading in importance (I think people are starting to spot them and name them more), so it feels like it is worth taking a little extra time to design this.

Reply:
Yes, sorry it hadn't been clear to me.

And given the importance of the concept (I agree: it's not just me ;-)), I think the overall effect of your comments is to nudge me towards putting this back into DRAFT until I can get a clearer picture of how best to proceed forwards. A  *lot* of food for thought (the perils of being the first to contribute a 'significant' addition, one which opens a new front in the hierarchy...?)

Reply:
> By 'named definition' I meant like `Commutative` and `Associative` -- maybe that wasn't clear? My feeling is that if we looked, we'd find uses of something equivalent to this action structure already in the library!

To be fair, I haven't looked in the library, but I am worried about where such a `Definition` should go... because `Algebra.Definitions` is single-sorted... and for the sake of the one definition, is it worth abstracting out? And where do you think it should go?

Reply:
On the last question: `Algebra.Action.Definitions`.